pkg/generate/code/equal.go

```diff
@@ -0,0 +1,236 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package code
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	awssdkmodel "github.com/aws/aws-sdk-go/private/model/api"
+
+	ackmodel "github.com/aws-controllers-k8s/code-generator/pkg/model"
+	"github.com/aws-controllers-k8s/code-generator/pkg/names"
+)
+
+// CheckEqualType returns Go code that checks the equality of two struct types.
+// The generated code returns false at the first spotted difference.
+func CheckEqualType(
+	typeDef *ackmodel.TypeDef,
+	// String representing the name of the variable that represents the first
+	// object under comparison. This will typically be something like "a"
+	firstVarName string,
+	// String representing the name of the variable that represents the second
+	// object under comparison. This will typically be something like "b".
+	secondVarName string,
+	// Number of levels of indentation to use
+	indentLevel int,
+) string {
+	out := ""
+
+	// We need a deterministic order to loop over attributes
+	attrsNames := []string{}
+	for attrName := range typeDef.Attrs {
+		attrsNames = append(attrsNames, attrName)
+	}
+	sort.Strings(attrsNames)
+
+	// For each attribute generate the equality check logic
+	for _, attrName := range attrsNames {
+		attr := typeDef.Attrs[attrName]
+		out += checkEqualField(attrName, attr.Shape, firstVarName, secondVarName, indentLevel)
+	}
+	return out
+}
+
+// checkEqualField outputs Go code that compares two similar fields from different objects.
+// The generated code first checks the nility of the fields and then procede on comparing
+// their value.
+func checkEqualField(
+	// fieldName is the field to generate the comparison logic for
+	fieldName string,
+	shape *awssdkmodel.Shape,
+	// String representing the name of the variable that represents the object
+	// under comparison.
+	firstVarName string,
+	// String representing the name of the variable that represents the second
+	// object under comparison.
+	secondVarName string,
+	// Number of levels of indentation to use
+	indentLevel int,
+) string {
+	out := ""
+
+	memberNames := names.New(fieldName)
+	memberNameClean := memberNames.Camel
+	firstAdaptedVarName := firstVarName + "." + memberNameClean
+	secondAdaptedVarName := secondVarName + "." + memberNameClean
+
+	nilCode := checkEqualNil(
+		shape,
+		firstAdaptedVarName,
+		secondAdaptedVarName,
+		indentLevel,
+	)
+	if nilCode != "" {
+		out += fmt.Sprintf(
+			"%s else if %s != nil && %s != nil {\n",
+			nilCode, firstAdaptedVarName, secondAdaptedVarName,
+		)
+		indentLevel++
+	} else {
+		out += "\n"
+	}
+
+	indent := strings.Repeat("\t", indentLevel)
+
+	switch shape.Type {
+	case "structure":
+		// We just re-use the generated functions to compare field of `struct` type
+		out += fmt.Sprintf(
+			"%sif !%s(%s, %s) {\n",
+			indent,
+			shape.ShapeName,
+			firstAdaptedVarName,
+			secondAdaptedVarName,
+		)
```

Comment:
go ahead and add another parameter to this function, called `comparePackageAlias` and add this variable into the output Go code here, so that if this `comparePackageAlias` parameter was `svccompare` and the Shape name was `ImageScanningConfiguration`, the output Go code would be:

```go
if !svccompare.IsEqualImageScanningConfiguration(a.ImageScanningConfiguration, b.ImageScanningConfiguration) {
```

Reply:
`IsEqualImageScanningConfiguration` is in this same package and doesn't need to be imported.
I think this is something we should do in `pkg/generate/code/compare.go` (code that generates `newResourceDelta`).

Reply:
> `IsEqualImageScanningConfiguration` is in this same package and doesn't need to be imported.

Ahhhh, yes, correct you are!

> I think this is something we should do in `pkg/generate/code/compare.go` (code that generates `newResourceDelta`).

:+1: